mantidproject/mantid
# Crystal Field - Allow arbitrary J (Ion type)

Changes the CrystalField code to allow non-rare earth ions (e.g. transition metal ions, assuming spin-only / orbitally quenched) to be calculated. For these ions, the `g_J` value is fixed to 2 (spin-only value).

**To test:**

Initialise a `CrystalField` object with `Ion=S<n>` where `<n>` is an integer or half-integer number (decimal notation, e.g. `Ion=S1.5`), and set `B20` to 1. Print out the Hamiltonian and check that it has `2<n>+1` rows/columns and the diagonal elements follow the sequence: `3*Jz^2-J(J+1)`, where the Jz runs from -J to +J. E.g.

```
J = 1
cf = CrystalField.fitting.CrystalField('J'+str(J), 'C2v', B20=1)
print cf.getHamiltonian()
checksum = np.sum(np.sum( np.diag(3*(np.arange(-J,J+1)**2) - J*(J+1)) - cf.getHamiltonian() ))
print "checksum = " + str(checksum)
```

should give the output:

```
[[ 1.+0.j  0.+0.j  0.+0.j]
 [ 0.+0.j -2.+0.j  0.+0.j]
 [ 0.+0.j  0.+0.j  1.+0.j]]
checksum = 0j
```

Check that checksum is always `0` for any value of `J`.

Also, check that all the usual operations work without errors, e.g.

```
cf = CrystalField('J'+str(J), 'C2v', B20=1, FWHM=0.1,Temperature=10)
spec = cf.getSpectrum();
plot(*spec)
plot(*cf.getSusceptibility(range(1,300), Hdir=[1, 1, 1], Lambda=-10))
ws = CreateWorkspace(*spec)
cf1 = CrystalField('J'+str(J), 'C2v', B20=0.9, FWHM=0.2,Temperature=10)
fit = CrystalFieldFit(Model=cf, InputWorkspace=ws, MaxIterations=100)
fit.fit()
```

Fixes #17984

In release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
